File: mcp-code-intelligence-python/src/mcp_code_intel/orchestration/registry/registry.py
"""Unified registry — merges native + child server tools into a searchable index.

Supports fallback chains, tokenized search, hit tracking, and session toggles.
Behavioral parity with Kotlin UnifiedRegistry.kt.
"""

from __future__ import annotations

from .grouper import ChainEntry, RegisteredTool, SemanticGrouper, ToolChain
from .tokenizer import tokenize

META_TOOL_NAMES = frozenset({
    "find_tools", "execute_dynamic_tool", "toggle_tool",
    "reset_tools", "manage_auto_approve", "orchestration_status", "agent_log",
})


class UnifiedRegistry:
    """Searchable tool index with fallback chains and hit-based ranking."""

    def __init__(self, similarity_threshold: float = 0.7) -> None:
        self._similarity_threshold = similarity_threshold
        self._native_tools: list[RegisteredTool] = []
        self._child_tools: list[RegisteredTool] = []
        self._merged: list[RegisteredTool] = []
        self._toggles: dict[str, bool] = {}
        self._chains: dict[str, ToolChain] = {}
        self._server_order: list[str] = []
        self._hits: dict[str, int] = {}

    def set_server_order(self, order: list[str]) -> None:
        """Set config-declared server order (determines fallback priority)."""
        self._server_order = order

    def set_child_tools(self, server_name: str, tools: list[dict]) -> None:
        """Register tools from a child server (filters meta-tools)."""
        filtered = [t for t in tools if t.get("name", "") not in META_TOOL_NAMES]
        priority = self._server_order.index(server_name) if server_name in self._server_order else 999
        self._child_tools = [t for t in self._child_tools if t.source != f"child:{server_name}"]
        for defn in filtered:
            name = defn.get("name", "unknown")
            desc = defn.get("description", "")
            self._child_tools.append(RegisteredTool(
                name=name, definition=defn, source=f"child:{server_name}",
                priority=priority, name_tokens=tokenize(name), desc_tokens=tokenize(desc),
            ))
        self._rebuild()

    def search(self, query: str) -> list[RegisteredTool]:
        """Tokenized search — scores by relevance + popularity."""
        terms = tokenize(query)
        if not terms:
            return [t for t in self._merged if self.is_enabled(t.name)]
        max_hits = max((self._hits.get(t.name, 0) for t in self._merged), default=1) or 1
        scored = []
        for tool in self._merged:
            if not self.is_enabled(tool.name):
                continue
            score = self._combined_score(tool, terms, max_hits)
            if score > 0.0:
                scored.append((tool, score))
        scored.sort(key=lambda x: x[1], reverse=True)
        return [t for t, _ in scored]

    def find(self, name: str) -> RegisteredTool | None:
        """Find tool by exact name."""
        for t in self._merged:
            if t.name == name and self.is_enabled(t.name):
                return t
        return None

    def get_chain(self, tool_name: str) -> ToolChain | None:
        """Get fallback chain for a tool name."""
        return self._chains.get(tool_name)

    def record_hit(self, tool_name: str, weight: int = 1) -> None:
        """Record successful execution hit. Triggers decay if > 1000."""
        self._hits[tool_name] = self._hits.get(tool_name, 0) + weight
        if self._hits[tool_name] > 1000:
            self._apply_decay(tool_name)

    def toggle(self, tool_name: str, enabled: bool) -> None:
        """Toggle tool on/off for this session."""
        self._toggles[tool_name] = enabled

    def reset_toggles(self) -> None:
        """Reset all session toggles."""
        self._toggles.clear()

    def is_enabled(self, tool_name: str) -> bool:
        """Check if tool is enabled."""
        return self._toggles.get(tool_name, True)

    def get_all(self) -> list[dict]:
        """Get all enabled tool definitions."""
        return [t.definition for t in self._merged if self.is_enabled(t.name)]

    def child_tools_by_server(self) -> dict[str, list[str]]:
        """Get child tool names grouped by server."""
        result: dict[str, list[str]] = {}
        for t in self._child_tools:
            result.setdefault(t.source, []).append(t.name)
        return result

    def all_child_tools(self) -> list[RegisteredTool]:
        """Get all child tools as flat list."""
        return list(self._child_tools)

    def register_nested(self, unique_name: str, server_name: str, definition: dict) -> None:
        """Register a tool discovered via nested find_tools delegation."""
        existing = self.find(unique_name)
        if existing:
            return
        name = definition.get("name", unique_name)
        desc = definition.get("description", "")
        priority = self._server_order.index(server_name) if server_name in self._server_order else 999
        tool = RegisteredTool(
            name=unique_name, definition=definition, source=f"nested:{server_name}",
            priority=priority, name_tokens=tokenize(name), desc_tokens=tokenize(desc),
        )
        self._child_tools.append(tool)
        self._merged.append(tool)

    def _combined_score(self, tool: RegisteredTool, terms: set[str], max_hits: int) -> float:
        """Combined relevance + popularity score."""
        relevance = self._score_against_terms(tool, terms)
        if relevance <= 0.0:
            return 0.0
        normalized_hits = self._hits.get(tool.name, 0) / max_hits
        return normalized_hits * 0.6 + relevance * 0.4

    def _score_against_terms(self, tool: RegisteredTool, query_terms: set[str]) -> float:
        """Score tool against query terms."""
        score = 0.0
        for term in query_terms:
            if term in tool.name_tokens:
                score += 2.0
            elif any(term in dt for dt in tool.desc_tokens):
                score += 1.0
        return score / (len(query_terms) * 2.0) if query_terms else 0.0

    def _apply_decay(self, trigger_tool: str) -> None:
        """Decay: subtract 500 from group, floor at -2000."""
        chain = self._chains.get(trigger_tool)
        if chain:
            group_names = {e.tool_name or chain.tool_name for e in chain.entries} | chain.similar_names
        else:
            group_names = {trigger_tool}
        for name in group_names:
            if name in self._hits:
                self._hits[name] = max(-2000, self._hits[name] - 500)

    def _rebuild(self) -> None:
        """Rebuild merged list and chains."""
        merged_map: dict[str, RegisteredTool] = {}
        for t in self._child_tools:
            merged_map[t.name] = t
        for t in self._native_tools:
            merged_map[t.name] = t
        self._merged = list(merged_map.values())
        grouper = SemanticGrouper(self._similarity_threshold)
        self._chains = grouper.build_chains(self._child_tools)

File: backend/mcp-code-intelligence-nodejs/src/orchestration/registry/registry.ts
/**
 * Unified registry — merges native + child server tools into a searchable index.
 * Supports fallback chains, tokenized search, hit tracking, and session toggles.
 * Behavioral parity with Kotlin UnifiedRegistry.kt.
 */

import { tokenize } from './tokenizer.js';
import { SemanticGrouper, RegisteredTool, ToolChain } from './grouper.js';

const META_TOOL_NAMES = new Set([
  'find_tools', 'execute_dynamic_tool', 'toggle_tool',
  'reset_tools', 'manage_auto_approve', 'orchestration_status', 'agent_log',
]);

export class UnifiedRegistry {
  private nativeTools: RegisteredTool[] = [];
  private childTools: RegisteredTool[] = [];
  private merged: RegisteredTool[] = [];
  private toggles = new Map<string, boolean>();
  private chains = new Map<string, ToolChain>();
  private serverOrder: string[] = [];
  private hits = new Map<string, number>();

  constructor(private similarityThreshold: number = 0.7) {}

  setServerOrder(order: string[]): void { this.serverOrder = order; }

  /** Register tools from a child server (filters meta-tools). */
  setChildTools(serverName: string, tools: Record<string, any>[]): void {
    const filtered = tools.filter((t) => !META_TOOL_NAMES.has(t.name ?? ''));
    const priority = this.serverOrder.indexOf(serverName);
    const prio = priority < 0 ? 999 : priority;
    this.childTools = this.childTools.filter((t) => t.source !== `child:${serverName}`);
    for (const defn of filtered) {
      const name = defn.name ?? 'unknown';
      const desc = defn.description ?? '';
      this.childTools.push({
        name, definition: defn, source: `child:${serverName}`,
        priority: prio, nameTokens: tokenize(name), descTokens: tokenize(desc),
      });
    }
    this.rebuild();
  }

  /** Tokenized search — scores by relevance + popularity. */
  search(query: string): RegisteredTool[] {
    const terms = tokenize(query);
    if (terms.size === 0) return this.merged.filter((t) => this.isEnabled(t.name));
    const maxHits = Math.max(1, ...this.merged.map((t) => this.hits.get(t.name) ?? 0));
    const scored: Array<[RegisteredTool, number]> = [];
    for (const tool of this.merged) {
      if (!this.isEnabled(tool.name)) continue;
      const score = this.combinedScore(tool, terms, maxHits);
      if (score > 0) scored.push([tool, score]);
    }
    scored.sort((a, b) => b[1] - a[1]);
    return scored.map(([t]) => t);
  }

  find(name: string): RegisteredTool | null {
    return this.merged.find((t) => t.name === name && this.isEnabled(t.name)) ?? null;
  }

  /** Compute deterministic hash of current tool registry state. */
  versionHash(): string {
    const names = this.merged.map((t) => t.name).sort().join('|');
    let hash = 0;
    for (let i = 0; i < names.length; i++) {
      hash = ((hash << 5) - hash + names.charCodeAt(i)) | 0;
    }
    return (hash >>> 0).toString(16).padStart(8, '0');
  }

  getChain(toolName: string): ToolChain | null { return this.chains.get(toolName) ?? null; }

  recordHit(toolName: string, weight: number = 1): void {
    const current = this.hits.get(toolName) ?? 0;
    this.hits.set(toolName, current + weight);
    if (current + weight > 1000) this.applyDecay(toolName);
  }

  toggle(toolName: string, enabled: boolean): void { this.toggles.set(toolName, enabled); }
  resetToggles(): void { this.toggles.clear(); }
  isEnabled(toolName: string): boolean { return this.toggles.get(toolName) ?? true; }

  getAll(): Record<string, any>[] { return this.merged.filter((t) => this.isEnabled(t.name)).map((t) => t.definition); }

  childToolsByServer(): Map<string, string[]> {
    const result = new Map<string, string[]>();
    for (const t of this.childTools) {
      const list = result.get(t.source) ?? [];
      list.push(t.name);
      result.set(t.source, list);
    }
    return result;
  }

  allChildTools(): RegisteredTool[] { return [...this.childTools]; }

  /** Register a tool discovered via nested find_tools delegation. */
  registerNested(uniqueName: string, serverName: string, definition: Record<string, any>): void {
    const desc = definition.description ?? '';
    const tool: RegisteredTool = {
      name: uniqueName,
      definition,
      source: `child:${serverName}`,
      priority: this.serverOrder.indexOf(serverName) < 0 ? 999 : this.serverOrder.indexOf(serverName),
      nameTokens: tokenize(uniqueName),
      descTokens: tokenize(desc),
    };
    this.childTools.push(tool);
    this.rebuild();
  }

  private combinedScore(tool: RegisteredTool, terms: Set<string>, maxHits: number): number {
    const relevance = this.scoreAgainstTerms(tool, terms);
    if (relevance <= 0) return 0;
    const normalizedHits = (this.hits.get(tool.name) ?? 0) / maxHits;
    return normalizedHits * 0.6 + relevance * 0.4;
  }

  private scoreAgainstTerms(tool: RegisteredTool, queryTerms: Set<string>): number {
    let score = 0;
    for (const term of queryTerms) {
      if (tool.nameTokens.has(term)) { score += 2.0; }
      else if ([...tool.descTokens].some((dt) => dt.includes(term))) { score += 1.0; }
    }
    return queryTerms.size > 0 ? score / (queryTerms.size * 2.0) : 0;
  }

  private applyDecay(triggerTool: string): void {
    const chain = this.chains.get(triggerTool);
    const groupNames = chain
      ? new Set([...chain.entries.map((e) => e.toolName ?? chain.toolName), ...chain.similarNames])
      : new Set([triggerTool]);
    for (const name of groupNames) {
      if (this.hits.has(name)) {
        this.hits.set(name, Math.max(-2000, (this.hits.get(name) ?? 0) - 500));
      }
    }
  }

  private rebuild(): void {
    const map = new Map<string, RegisteredTool>();
    for (const t of this.childTools) map.set(t.name, t);
    for (const t of this.nativeTools) map.set(t.name, t);
    this.merged = [...map.values()];
    const grouper = new SemanticGrouper(this.similarityThreshold);
    this.chains = grouper.buildChains(this.childTools);
  }
}
